controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1041 @@
+		fs, err := s.FormationList(app.ID)
+		if err != nil {
+			log.Error("error getting formations", "err", err)
+			continue
+		}
+		log.Info(fmt.Sprintf("got %d formation(s) for %s app", len(fs), app.Name))
+
+		for _, f := range fs {
+			log.Info("updating formation", "release.id", f.ReleaseID)
+			if _, err := s.updateFormation(f); err != nil {
+				log.Error("error updating formation", "release.id", f.ReleaseID, "err", err)
+			}
+		}
+	}
+
+	s.triggerRectify()
+}
+
+func (s *Scheduler) Rectify() {
+	log := logger.New("fn", "Rectify")
+
+	if !s.isLeader {
+		log.Warn("ignoring rectify as not service leader")
+		return
+	}
+
+	defer s.sendEvent(NewEvent(EventTypeRectify, nil, nil))
+
+	log.Info("starting rectify")
+
+	pending := NewPendingJobs(s.jobs)
+	pending.Update(s.pendingStarts)
+
+	for key := range pending {
+		formationLog := log.New("app.id", key.AppID, "release.id", key.ReleaseID)
+		formationLog.Info("rectifying formation")
+		formation, ok := s.formations[key]
+		if !ok {
+			formationLog.Info("unknown formation, getting from the controller")
+			cf, err := s.GetFormation(key.AppID, key.ReleaseID)
+			if err != nil {
+				formationLog.Error("error getting formation", "err", err)
+				continue
+			}
+			formation, err = s.updateFormation(cf)
+			if err != nil {
+				formationLog.Error("error updating formation", "err", err)
+				continue
+			}
+		}
+
+		expected := formation.Processes
+		actual := pending.GetProcesses(key)
+
+		if len(actual) == 0 && len(expected) == 0 || reflect.DeepEqual(actual, expected) {
+			formationLog.Info("formation in correct state", "expected", expected, "actual", actual)
+			continue
+		}
+		// TODO: make this formation.Diff(actual)?
+		formation.Processes = actual
+		diff := formation.Update(expected)
+		formationLog.Info("formation in incorrect state", "expected", expected, "actual", actual, "diff", diff)
+		s.sendDiffRequests(formation, diff)
+	}
+
+	for key, formation := range s.formations {
+		if _, ok := pending[key]; !ok && len(formation.Processes) > 0 {
+			log = log.New("app.id", key.AppID, "release.id", key.ReleaseID)
+			log.Info("formation in incorrect state", "expected", formation.Processes, "actual", nil, "diff", formation.Processes)
+			s.sendDiffRequests(formation, formation.Processes)
+		}
+	}
+}
+
+func (s *Scheduler) HandleFormationChange(ef *ct.ExpandedFormation) {
+	var err error
+	defer func() {
+		s.sendEvent(NewEvent(EventTypeFormationChange, err, nil))
+	}()
+
+	log := logger.New("fn", "HandleFormationChange")
+	if ef.App != nil {
+		log = log.New("app.id", ef.App.ID)
+	}
+	if ef.Release != nil {
+		log = log.New("release.id", ef.Release.ID)
+	}
+	log.Info("handling formation change")
+	_, err = s.changeFormation(ef)
+	if err != nil {
+		log.Error("error handling formation change", "err", err)
+		return
+	}
+
+	// Trigger sync jobs in case we've ignored an existing job because we
+	// didn't know about the formation.
+	s.triggerSyncJobs()
```

Comment:
When we get a job event, we [ask the controller for the formation if we don't already have it](https://github.com/flynn/flynn/blob/scheduler-v2-review/controller/scheduler-v2/scheduler.go#L580-L592), so do we need to trigger a job sync after each formation change, rather than just a rectify?

My main thinking here is that `SyncJobs` does quite a lot of work ([looks up all the hosts](https://github.com/flynn/flynn/blob/scheduler-v2-review/controller/scheduler-v2/scheduler.go#L212), [gets the jobs from each host](https://github.com/flynn/flynn/blob/scheduler-v2-review/controller/scheduler-v2/scheduler.go#L223), [saves the jobs](https://github.com/flynn/flynn/blob/scheduler-v2-review/controller/scheduler-v2/scheduler.go#L596) which leads to both [handleJobStart](https://github.com/flynn/flynn/blob/scheduler-v2-review/controller/scheduler-v2/scheduler.go#L814) and [putJobs](https://github.com/flynn/flynn/blob/scheduler-v2-review/controller/scheduler-v2/scheduler.go#L820) being called for every active job).